src/coreclr/jit/lower.cpp

```diff
@@ -7004,6 +7022,181 @@ void Lowering::ContainCheckJTrue(GenTreeOp* node)
+    {
+        if (child->OperIs(GT_AND))
+        {
+            // If Op2 is not contained, then try to contain it.
+            if (!child->AsOp()->gtGetOp2()->isContainedAndNotIntOrIImmed())
+            {
+                if (!ContainCheckCompareChain(child->gtGetOp2(), child, startOfChain))
+                {
+                    // Op2 must be contained in order to contain Op1 or the AND.
+                    return false;
+                }
+            }
+
+            // If Op1 is not contained, then try to contain it.
+            if (!child->AsOp()->gtGetOp1()->isContainedAndNotIntOrIImmed())
+            {
+                if (!ContainCheckCompareChain(child->gtGetOp1(), child, startOfChain))
+                {
+                    return false;
+                }
+            }
+
+            // Contain the AND.
+            child->SetContained();
+            return true;
+        }
+#ifdef TARGET_ARM64
+        else if (child->OperIsCompare())
+        {
+            child->AsOp()->SetContained();
+
+            // Ensure the children of the compare are contained correctly.
+            child->AsOp()->gtGetOp1()->ClearContained();
+            child->AsOp()->gtGetOp2()->ClearContained();
+            ContainCheckConditionalCompare(child->AsOp());
+            *startOfChain = child;
+            return true;
+        }
+#endif
+    }
+
+    return false;
+}
+
+//------------------------------------------------------------------------
+// ContainCheckAndCompareChain : Determine if an AND is a containable chain
+//
+// Arguments:
+//    node - pointer to the node
+//
+void Lowering::ContainCheckAndChain(GenTree* tree)
+{
+    assert(tree->OperIs(GT_AND));
+
+    if (!comp->opts.OptimizationEnabled())
+    {
+        return;
+    }
+
+    // First check there is a valid chain.
+    if (IsValidCompareChain(tree->AsOp()->gtGetOp2(), tree) &&
+        IsValidCompareChain(tree->AsOp()->gtGetOp1(), tree))
+    {
+        GenTree* startOfChain = nullptr;
+
+        // To ensure ordering at code generation, Op1 and the parent can
+        // only be contained if Op2 is contained.
+        if (ContainCheckCompareChain(tree->AsOp()->gtGetOp2(), tree, &startOfChain))
+        {
+            if (ContainCheckCompareChain(tree->AsOp()->gtGetOp1(), tree, &startOfChain))
+            {
+                // If op1 is the start of a chain, then it'll be generated as a standard compare.
+                if (startOfChain != nullptr)
+                {
+                    // The earliest node in the chain will be generated as a standard compare.
+                    assert(startOfChain->OperIsCompare());
+                    startOfChain->AsOp()->gtGetOp1()->ClearContained();
+                    startOfChain->AsOp()->gtGetOp2()->ClearContained();
+                    ContainCheckCompare(startOfChain->AsOp());
+                }
+            }
+        }
+
+        JITDUMP("Lowered And chain:\n");
+        DISPTREE(tree);
+    }
+}
```

Comment:
`IsValidCompareChain`, `ContainCheckCompareChain` and `ContainCheckAndChain` should go in lowerarmarch.cpp an be conditioned on arm64.